include/eosio/vm/backend.hpp

```diff
@@ -74,49 +76,68 @@ namespace eosio { namespace vm {
+         : memory_alloc(alloc), ctx(new context_t{parse_module(code, options), detail::get_max_call_depth(options)}) {
+         ctx->set_max_pages(detail::get_max_pages(options));
          construct(&host);
       }
       backend(wasm_code&& code, wasm_allocator* alloc, const Options& options = Options{})
-         : memory_alloc(alloc), ctx(parse_module(code, options), detail::get_max_call_depth(options)) {
-         ctx.set_max_pages(detail::get_max_pages(options));
+         : memory_alloc(alloc), ctx(new context_t{parse_module(code, options), detail::get_max_call_depth(options)}) {
+         ctx->set_max_pages(detail::get_max_pages(options));
          construct();
       }
       backend(wasm_code& code, host_t& host, wasm_allocator* alloc, const Options& options = Options{})
-         : memory_alloc(alloc), ctx(parse_module(code, options), detail::get_max_call_depth(options)) {
-         ctx.set_max_pages(detail::get_max_pages(options));
+         : memory_alloc(alloc), ctx(new context_t{parse_module(code, options), detail::get_max_call_depth(options)}) {
+         ctx->set_max_pages(detail::get_max_pages(options));
          construct(&host);
       }
       backend(wasm_code& code, wasm_allocator* alloc, const Options& options = Options{})
-         : memory_alloc(alloc), ctx(parse_module(code, options), detail::get_max_call_depth(options)) {
-         ctx.set_max_pages(detail::get_max_pages(options));
+         : memory_alloc(alloc), ctx(new context_t{(parse_module(code, options)), detail::get_max_call_depth(options)}) {
+         ctx->set_max_pages(detail::get_max_pages(options));
          construct();
       }
       backend(wasm_code_ptr& ptr, size_t sz, host_t& host, wasm_allocator* alloc, const Options& options = Options{})
-         : memory_alloc(alloc), ctx(parse_module2(ptr, sz, options, true), detail::get_max_call_depth(options)) { // single parsing. original behavior
-         ctx.set_max_pages(detail::get_max_pages(options));
+         : memory_alloc(alloc), ctx(new context_t{parse_module2(ptr, sz, options, true), detail::get_max_call_depth(options)}) { // single parsing. original behavior {
+         ctx->set_max_pages(detail::get_max_pages(options));
          construct(&host);
       }
       // Leap:
       //  * Contract validation only needs single parsing as the instantiated module is not cached.
       //  * JIT execution needs single parsing only.
       //  * Interpreter execution requires two-passes parsing to prevent memory mappings exhaustion
-      backend(wasm_code_ptr& ptr, size_t sz, wasm_allocator* alloc, const Options& options = Options{}, bool single_parsing = true)
-         : memory_alloc(alloc), ctx(parse_module2(ptr, sz, options, single_parsing), detail::get_max_call_depth(options)) {
-         ctx.set_max_pages(detail::get_max_pages(options));
+      //  * Leap reuses execution context per thread; is_exec_ctx_created_by_backend is set
+      //  to false when a backend is constructued
```

Comment:
As far as I know Leap is the only user of this submodule. Seems to me we should just change this to decouple exec ctx from backend. Supporting both modes, one of which is not used seems pointless.

Reply:
Thanks Kevin and Matt for this. I see `eos-vm` is submoduled in CDT `tools/external`. We can remove the dual modes (and make the code simpler) once decision on CDT is made.

Reply:
Ok, sounds good.